rts/motoko-rts/src/bigint.rs

```diff
@@ -250,6 +250,9 @@ unsafe extern "C" fn bigint_of_int64(j: i64) -> Value {
 #[cfg(feature = "ic")]
 #[no_mangle]
 unsafe extern "C" fn bigint_of_float64(j: f64) -> Value {
+    // handle fast path: these numbers (when rounded towards zero)
+    // can be represented as `Int` without resorting to heap allocation,
+    // i.e. >= 0xc0000000 and <= 0x3fffffff
```

Comment:
Okay, here is my expanded reasoning:
- I know that for `n >= 0xc0000000 and n <= 0x3fffffff` with `n : Int` we can have compact bignums (cheap, non-heap-allocated numbers that fit into a `Vanilla`-represented value
- if you write those numbers as decimal, you'll see `-1073741824` and `1073741823`
- now deduce backwards which `Float` range maps to it under the Rust operation `(j: f64) as i32`
- it turns out that `-1073741825.0 < f < 1073741824.0 is this range, since `as i32` rounds towards zero

Reply:
Thanks, this answers my question. Could we copy this explanation to code? Or maybe something like:

```rust
// BigInts in range [-1073741824 (0xc0000000), 1073741823 (0x3fffffff)] are
// represented as scalars
if  (j as i32) >= -1073741824 && (j as i32) <= 1073741823 {
    ...
}
```

Note that I also refactored the code a little bit above. First, instead of using `... < 1073741824` I used `... <= 1073741823` (and same for the other condition) to make sure the constants in code and comments match. Secondly, I used `(j as i32) >= -1073741824` instead of `j >= -1073741824.0` (and same for the other condition). This seems more direct because in the body of the conditional we rely on truncation of the float to be in the described (in the comments) range.

This makes the code more clear to me, but that's a personal thing (others might find it harder to understand) so no need to incorporate any of it. Alternatively, I could submit a PR with these changes.

Reply:
Does Rust not support hex float literals? That would make this more obvious.

Reply:
Rust doesn't have hex or binary float literals ([source](https://doc.rust-lang.org/reference/tokens.html#floating-point-literals)).

Since we care about whether the truncated version is in the range, I think it makes sense to truncate first, then check the truncated version (integer).

Reply:
While it wasn't my aim to do refactoring (just adding comments), I am happy to go with such a rewrite. My only fear is that doing the conversion before comparison will make the slow path more expensive. Also you seem to rely on CSE to eliminate the second truncation operation.

Reply:
Then there is another problem. Truncation is not total with a target type of `i32`. This can lead to trouble.

Reply:
Looks like the truncation rules use saturated arithmetic (in `lldb`) so maybe there is no problem at all:
```
(lldb) p/x (int32_t) -6677.9e140
(int32_t) $7 = 0x80000000
(lldb) p/x (int32_t) 6677.9e140
(int32_t) $8 = 0x7fffffff
```
If Rust does the same your rewrite would be safe. Still I don't prefer it to the float comparisons. I'll only change the comments, but have no objections if you change the code in a subsequent PR :-)

Reply:
OK, thanks.

Just to clarify my point, we can store *integers* in that range without allocation, not *floats*. So IMO it makes more sense to check the *truncated float*, rather than the float itself, because we care about whether the truncated version is in the range.

Reply:
Sure I am saying that with "represented as `Int`". I am also saying "rounded towards zero". By using float comparisons I also sidestep the issue of infinities and `NaN`s (because the condition is known to result in `false`), so my fast path is a conservative optimisation to what `mp_float_to_int` does.

Reply:
> I am saying that with "represented as Int"

You're saying that in comments, not in code. Code does something different :-)

> By using float comparisons I also sidestep the issue of infinities and NaNs

OK, this makes sense. I just tried this, it seems like in `f64::NAN as i32` is 0, but `mp_set_double(..., NAN)` fails. So if we truncate and check the truncated number, we need to check the `NAN` case (and probably INF cases too) first and have an early return path returning an error.

Reply:
`Value::from_signed_scalar` is doing that. It gets passed a `i32` and does the right thing. How does it something different?